File: quick_sort.cpp
/*
* @File      :   quick_sort.cpp
* @Date      :   2021/10/02 18:07:12
* @Author    :   Quiser
* @Email     :   qs0714@foxmail.com
*/
#include <iostream>
#include <vector>
#include <ctime>
#define random(x, y) (rand() % (y - x) + x)
using namespace std;


void showvec(vector<int> &l)
{
    for (auto &&i : l)
    {
        cout << i << " ";
    }
    cout << endl;
}


void quick_sort(vector<int> &l, int left, int right)
{
    if (right <= left)
        return;
    // for (int i = left; i < right; i++)
    // {
    //     if (l[i + 1] < l[i])
    //         break;
    //     if (i == right - 1)
    //         return;
    // }
    int key = l[left];
    int mark = left;
    for (int i = left+1; i <= right; i++)
    {
        if (l[i] <= key)
        {
            mark++;
            swap(l[i], l[mark]);
        }
    }
    swap(l[left], l[mark]);
    showvec(l);
    quick_sort(l, left, mark - 1);
    quick_sort(l, mark + 1, right);
}

int main()
{
    cout << "start" << endl;
    srand((int)time(0));
    // for (int  i = 0; i < 100; i++)
    // {
    //     cout<<random(5,10)<<" ";
    // }
    // cout<<endl;

    vector<int> l = {};
    for (int i = 0; i < 50; i++)
    {
        l.push_back(random(1,50));
    }

    showvec(l);
    quick_sort(l, 0, l.size() - 1);
    showvec(l);
    cout << "end" << endl;
    return 0;
}

File: quick_sort.py
#!/usr/bin/env python3

from random import randint

def qsort(l,left,right):
    if right - left <= 0:
        return
    for i in range(left,right):
        if l[i+1]<l[i]:
            break
        if i == right-1:
            return
    key = l[randint(left,right)]
    i=left
    j=right
    while True:
        while i < j and l[i] < key:
            i+=1
        while i < j and l[j] >= key:
            j-=1
        if j <= i:
            break
        l[i],l[j]=l[j],l[i]
    #print(left,right,key,i,j,l)
    qsort(l,left,i-1)
    qsort(l,i,right)
l = [randint(1,200) for i in range(200)]
print(l)
qsort(l,0,len(l)-1)
print(l)

